---
title: "class15: Pathway Analysis from RNA-Seq Results"
author: "Duy Tong"
date: "11/14/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## About our Input Data
The data for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

>Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

## Import Count and Metadata

```{r}
metaFile <-"GSE37704_metadata.csv"
countFile <-"GSE37704_featurecounts.csv"

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```

```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
#ENSG = ENSEMBLES GENES
```

Dont match because there is a length data, so remove the column. The *length* column in countData is going to cause problems with the required atching to the metadata file so lets remove it here.

```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
# Adding the - sign meaning to remove it
```

Double Check that colnames in countData match 

```{r}
colnames(countData)
```

Access SRR. We can use the all() function to check if all entries in a vector are TRUE
```{r}
all(colnames(countData) == rownames(colData))
```

```{r}
# Test how the all() function works
all(c(T,F,T))
```

## Remove Zero Count Genes

We want to remove genes that have 0 count values in all experiments (i.e. rows that have 0 accross all cols).

```{r}
# == equals to , != not equals to
# Filter out the 0
countData = countData[rowSums(countData) != 0,  ]
head(countData)
```

# DESeq2 Analysis
```{r}
library(DESeq2)
```

```{r}
# Setup the Object With Our Data in the Way of DESeq wants it
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
#Run the Analysis
dds = DESeq(dds)
```

Get Our Results
```{r}
res = results(dds)
```
```{r}
res
```

Summary
```{r}
summary(res)
```

```{r}
plot(res$log2FoldChange, -log(res$padj))
```

lets add some color
```{r}
mycols <- rep("gray", length(res$padj))
mycols[abs(res$log2FoldChange)>2] <- "blue"
mycols[res$padj >0.005] <- "gray"
plot(res$log2FoldChange, -log(res$padj), col = mycols)
```

## Add Gene Symbols and Entrez ids
```{r eval=FALSE}
BiocManager::install("AnnotationDbi")
BiocManager::install("org.Hs.eg.db")
```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

# We can translate between all the following database ID systems
columns(org.Hs.eg.db)
```

```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys= row.names(countData), # where are my IDs 
                    keytype="ENSEMBL",          # what format are my IDs
                    column= "SYMBOL",           # the new format I want
                    multiVals="first")
res
```

```{r}
res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(countData),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")
res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")
head(res,10)
```


# Section #2 Pathway Analysis
Install packages

Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)    #Human signaling and metabolism datasets
data(sigmet.idx.hs)   #

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

The main gage() function requires a named vector of fold changes, where the names of the values are the Entrez gene IDs.

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
#Get the results
keggres = gage(foldchanges, gset=kegg.sets.hs)
```

```{r}
#Look at Object Returned from gage()
attributes(keggres)
```

```{r}
#Look at the first few down (less) pathways
head(keggres$less)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
# gray boxes don't show any changes.
```

```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

Finally, lets pass these IDs in keggresids to the pathview() function to draw plots for all the top 5 pathways.
```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

>Q.Can you do the same procedure as above to plot the pathview figures for the top 5 down-reguled pathways?

# Section #3: Gene Ontology (GO)

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```











