---
title: "class14: Genome Informatics II"
author: "Duy Tong"
date: "11/12/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Background Information

  Data from a published RNA-seq experiement where airway smooth muscle cells were treated with dexamethasone, synthetic glucocorticoid steroid with anti-inflammatory effects.
  Glucocorticoids can be used to treat patients with asthma to reduce inflammation of the airways.

# Section #1: Bioconductor and DESeq2 Setup
```{r eval=FALSE}
install.packages("BiocManager")
BiocManager::install("DESeq2")

#Can check whether or not you have installed... Check in console
#library(BiocManager)
#library(DESeq2)

#Always install in console!
```
 
Side-Notes: Aligning Reads to Reference Genome

The computational analysis of an RNA-seq experiment begins from the FASTQ files that contain the nucleotide sequence of each read and a quality score at each position. These reads must first be aligned to a reference genome or transcriptome. The output of this alignment step is commonly stored in a file format called SAM/BAM. 

Once the reads have been aligned, there are a number of tools that can be used to count the number of reads/fragments that can be assigned to genomic features for each sample. These often take as input SAM/BAM alignment files and a file specifying the genomic features, e.g. a GFF3 or GTF file specifying the gene models as obtained from **ENSEMBLE** or **UCSC**.

DESeq2 Required Inputs

countData: count matrix; number of reads mapping to each gene for each sample.

colData: describes metadata about the columns of countData.

**First column of colData MUST MATCH column names of countData**
 
# Section #2: Import countData and colData

```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("airway_metadata.csv", stringsAsFactors = FALSE)

#stringAsFactors: argument to the data.frame(), that indicates whether strings in a data frame should be treated as factor variables or as just plain strings.

#stringAsFactors = FALSE, meaning that you're planning to change the type of strings.
```


Take a look at the data.
```{r}
head(counts)
```

```{r}
head(metadata)
```

The counts data is showing each genes of each samples, ENSG are the genes. The metadata (colData) view is showing the the groups (i.e SRR1039508...). Just think that countData has more specific details whereas, the colData just shows only the columns details.

```{r}
#How many experiemnts do we have
nrow(counts)
ncol(counts)
nrow(metadata)
ncol(metadata)
#Total of 38694 genes and 9 experiments
```

#Section #3: Toy Differntial Gene Expression

We want to know if there is a difference in expression values for control (non-drug) vs. treated (drug added)

First step is to find which experienments were the control experiments and then get the average values across all control expeirments (or treated)

For Control Group.
```{r}
# == exactly equal to
control <- metadata[metadata[,"dex"]=="control",]
control$id
```

Now calculate the mean values across these control columns of **countdata**...
```{r}
# These are the rows for control groups
control.mean <- rowSums(counts[,control$id])/length(control$id)
names(control.mean) <- counts$ensgene
```

For the treatment group...
```{r}
treated <-metadata[metadata[,"dex"]=="treated",]
treated
```

Now calculate the mean values across these control columns of **countdata**
```{r}
treated.mean <- rowSums(counts[,treated$id])/length(treated$id)
names(control.mean) <- counts$ensgene
```

> Q1. How would you make the above code more robust? What would happen if you were to add more samples. Would the values obtained with the exact code above be correct?
  If more samples are added, then the length would change therefore the mean will increase.No, the code above will be different because more samples are added.

> Q2. Follow the same procedure for the treated samples (i.e. calculate the mean per gene accross drug treated samples and assign to a labeled vector called treated.mean)
  Already assigned.
  
Combine our Meancount Data for Bookeping Purposes.
```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

Create Scatter Plot to Plot Control Vs. Treated
```{r}
plot(meancounts$control.mean, meancounts$treated.mean, log = "xy", xlab = " log Control", ylab = " log Treated", main = "Control Vs. Treated")
```

Calculate log2foldchange

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

```{r}
#Exclude all the 0 from data.set only focus on changes
# == 0 means that which ones are 0.
# arr.ind = shows where and which columns and rows where values are 0.
zero.vals <-which(meancounts[,1:2]==0, arr.ind=TRUE)

# unique syntax removes any row with 0 values that are duplicated.
to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let’s filter the dataset both ways to see how many genes are up or down-regulated.

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

Calculate how many genes are upregulated...
```{r}
sum(up.ind)
```

Calculate how many genes are downregulated...
```{r}
sum(down.ind)
```

```{r}
head( mycounts[up.ind,])
```

## Skipped Section 4

## Section 5: DESeq2 Analysis
Let’s do this the right way. DESeq2 is an R package for analyzing count-based NGS data like RNA-seq. It is available from Bioconductor. Bioconductor is a project to provide tools for analyzing high-throughput genomic data including RNA-seq, ChIP-seq and arrays.

```{r}
library(DESeq2)
```

We will use the DESeqDataSetFromMatrix() function to build the required DESeqDataSet object and call it dds, short for our DESeqDataSet. If you get a warning about “some variables in design formula are characters, converting to factors” don’t worry about it. Take a look at the dds object once you create it.

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts, colData = metadata, design = ~dex, tidy = TRUE)
dds
```

Run DESeq2

```{r}
dds <-DESeq(dds)
```

Getting the Results
```{r}
res <- results(dds)
res
```

Summarize Data
```{r}
summary(dds)
```

## Summary Plot: Volcano Plot
Volcano Plot: further away from 0 -> bigger change. This figure will combine both Fold Change and the pv-value into one overivew figure indicating the porportion of genes with large scale signficant differences in their expression.

Volcano Plot: Black and White
```{r}
plot(res$log2FoldChange, -log(res$padj))
```

Volcano Plot With Colors...
```{r}
# Add Colors
mycols <- rep("gray", nrow(res))
# Make points with +2 -2 fold change blue
mycols[abs(res$log2FoldChange) > 2] <- "blue"
#make point below p-value cutoff gray
mycols[abs(res$padj) > 0.05] <- "gray"

plot(res$log2FoldChange, -log(res$padj), col = mycols)
#Adding abline
abline(v = c(-2,2), col="gray", lty =2)
abline(h = -log(0.05), col ="gray", lty =2)
```

```{r}
# Setup your point color vector
mycols <- rep("gray", nrow(res))
mycols [abs(res$log2FoldChange) >2] <-"red"

inds <-(res$padj < 0.01) & (abs(res$log2FoldChange) >2)
mycols[inds] <- "blue"
```

Save our results for next day...
```{r}
write.csv(res, file="expression_results.csv")
```
